File: PYTHON/BinaryCodeChecker.py
"""
1461. Check If a String Contains All Binary Codes of Size K

Return True if every binary code of length k appears
as a substring of s.
"""

class BinaryCodeChecker:

    def hasAllCodes(self, s: str, k: int) -> bool:

        n = len(s)
        total_codes = 1 << k

        # If string too short → impossible
        if n < total_codes + k - 1:
            return False

        seen = [False] * total_codes
        count = 0

        hash_val = 0
        mask = total_codes - 1

        for i in range(n):

            # Rolling hash of last k bits
            hash_val = ((hash_val << 1) & mask) | int(s[i])

            # Start checking once window size reaches k
            if i >= k - 1 and not seen[hash_val]:
                seen[hash_val] = True
                count += 1

                if count == total_codes:
                    return True

        return False


# ---------- USER INPUT DRIVER ----------
if __name__ == "__main__":
    s = input("Enter binary string: ")
    k = int(input("Enter k: "))

    solver = BinaryCodeChecker()
    result = solver.hasAllCodes(s, k)

    print("Contains all binary codes of size k:", result)

File: JAVA/Leetcode/BinaryCodeChecker.java
/*
1461. Check If a String Contains All Binary Codes of Size K

Given a binary string s and an integer k,
return true if every binary code of length k exists as a substring of s.

Total possible binary codes of length k = 2^k.

Examples:
Input: s = "00110110", k = 2 → Output: true
Input: s = "0110", k = 1 → Output: true
Input: s = "0110", k = 2 → Output: false

Constraints:
1 <= s.length <= 5 * 10^5
s contains only '0' and '1'
1 <= k <= 20
*/

import java.util.*;

class BinaryCodeChecker {

    public boolean hasAllCodes(String s, int k) {

        int n = s.length();

        // Total number of binary codes of size k
        int totalCodes = 1 << k;

        // If string is too short, impossible to contain all codes
        if (n < totalCodes + k - 1) {
            return false;
        }

        boolean[] seen = new boolean[totalCodes];
        int count = 0;

        int hash = 0;
        int mask = totalCodes - 1;

        byte[] bytes = s.getBytes();

        for (int i = 0; i < n; i++) {

            // I shift left and add new bit → rolling hash of k bits
            hash = ((hash << 1) & mask) | (bytes[i] - '0');

            // Once window size reaches k, I check if code is new
            if (i >= k - 1 && !seen[hash]) {
                seen[hash] = true;
                count++;

                // If all possible codes are found → done
                if (count == totalCodes) {
                    return true;
                }
            }
        }

        return false;
    }

    // ---------- USER INPUT DRIVER ----------
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter binary string: ");
        String s = sc.nextLine();

        System.out.print("Enter k: ");
        int k = sc.nextInt();

        BinaryCodeChecker solver = new BinaryCodeChecker();
        boolean result = solver.hasAllCodes(s, k);

        System.out.println("Contains all binary codes of size k: " + result);

        sc.close();
    }
}